src/CasProcessor.php

```diff
@@ -0,0 +1,97 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Drupal\oe_authentication;
+
+/**
+ * Helper functions to process Cas attributes.
+ */
+class CasProcessor {
+
+  const USER_CAS_ATTRIBUTE_MAPPING = [
+    'mail' => 'email',
+    'field_oe_firstname' => 'firstName',
+    'field_oe_lastname' => 'lastName',
+    'field_oe_department' => 'departmentNumber',
+    'field_oe_organisation' => 'domain',
+  ];
+
+  /**
+   * Converts an array of cas attributes into an array of drupal field/values.
+   *
+   * @param array $attributes
+   *   An array containing a series of cas attributes.
+   *
+   * @return array
+   *   An array containing a series of Drupal field names and values.
+   */
+  public static function convertCasAttributesToFieldValues(array $attributes): array {
+    $values = [];
+    foreach (static::USER_CAS_ATTRIBUTE_MAPPING as $field_name => $property_name) {
+      if (!empty($attributes[$property_name])) {
+        $values[$field_name] = $attributes[$property_name];
+      }
+    }
+    return $values;
+  }
+
+  /**
+   * Parses the EU Login attributes from the validation response.
+   *
+   * @param string $source
+   *   The string containing the validation response.
+   *
+   * @return array
+   *   An array containing the parsed attributes.
+   */
+  public static function processValidationResponseAttributes(string $source): array {
+    $dom = new \DOMDocument();
+    $dom->preserveWhiteSpace = FALSE;
+    $dom->encoding = "utf-8";
+
+    // Suppress errors from this function, as we intend to allow other
+    // event subscribers to work on the data.
+    if (@$dom->loadXML($source) === FALSE) {
```

Comment:
I'm not 100% sure on this section, in the context of the event subscriber this made sense, but in this class we can reasonably expect to be passed in a valid processable string.

I would suggest to provide an additional static method `isValidationResponse(string $response): bool` that contains this check and the `authenticationSuccess` one. Then this can be called from the event subscriber.

This would also make it clean, having the validation and processing in separate, reusable methods :heart_eyes:

Reply:
And on top of that, suppressing any potential error with '@' is not a practice that we want to use here I guess...

Reply:
I agree that suppressing errors is a bad practice, but the DOM extension is a very special case and here it is fine I think. The DOM extension is not a "native" PHP extension but it is wrapping the `libxml` library which is very chatty about errors when parsing, and the extension just bubbles up anything reported by `libxml`, even if the document ends up to be parsed adequately. It would just cause noise in our logs and possibly on the user's screen.

Suppressing the errors but checking the output and throwing an exception if anything goes wrong would also allow us to track any problems, but in a way that is not needlessly chatty about XML specifics :)